src/main/java/org/opentripplanner/routing/algorithm/raptor/transit/mappers/TransitLayerUpdater.java

```diff
@@ -23,15 +24,17 @@
 import static org.opentripplanner.routing.algorithm.raptor.transit.mappers.TripPatternMapper.mapOldTripPatternToRaptorTripPattern;
 
 /**
- * Update the TransitLayer from a set of TimeTables. TripPatterns are matched on id and replaced
- * by their updated versions. A list of TripPatternsForDate is copied from the TransitLayer for
- * each relevant date, updated and then atomically replaced in the TransitLayer.
+ * Update the TransitLayer from a set of TimeTables. A shallow copy is made of the TransitLayer
+ * (this also includes a shallow copy of the TripPatternsForDate map). TripPatterns are matched on
+ * id and replaced by their updated versions. The realtime TransitLayer is then switched out
+ * with the updated copy in an atomic operation. This ensures that any TransitLayer that is
+ * referenced from the Graph is never changed.
```

Comment:
The operation is atomic only in the sense that reference assignment is atomic in Java. Readers may already take for granted the low-level fact that pointer read-tearing cannot happen. What is important here is not so much that the reference write is atomic, but that the object is only "published" and visible to other threads after all writes are complete, and that the combined act of writing and publishing happens-before any reads.  Many people will associate the term "atomic" with this kind of update operations composed of two or more distinct actions. 

Unfortunately, although the reference assignment is atomic, I believe this whole update operation is not atomic, in the sense that if thread A writes to fields of object X before it shares a reference to X, thread Y may see the reference update before it sees the fields within the referenced object (and so on recursively into collections). In practice, in my experience I've never encountered outright failures with the approach of simply setting fields before revealing a reference to their containing object, but if I'm not mistaken the Java Memory Model does not guarantee this will work. Strictly speaking, the reading thread could see a partially initialized Map or even a null field. I will do some more research to design a truly correct mechanism for a single writer thread "publishing" completed objects to a pool of reader threads. That mechanism will be reused many times in different places, so it's good to perfect it. I think the main idea is only accessing the published reference through synchronized blocks, and setting final fields in a constructor may also be an important ingredient.

Reply:
Based on your comments and discussion with @t2gran I made the realtimeTransitLayer private, wrapped it in ConcurrentPublished<T> and made is accessible through a getter and setter.